mobile/test/common/mocks/common/mocks.h

```diff
@@ -0,0 +1,49 @@
+#include "gmock/gmock.h"
+#include "library/common/common/system_helper.h"
+
+namespace Envoy {
+namespace test {
+
+// Mock implementation of SystemHelper.
+class MockSystemHelper : public SystemHelper {
+public:
+  MockSystemHelper();
+
+  // SystemHelper:
+  MOCK_METHOD(bool, isCleartextPermitted, (absl::string_view hostname));
+};
+
+// SystemHelperPeer allows the replacement of the SystemHelper singleton
+// with a MockSystemHelper.
+class SystemHelperPeer {
+public:
+  class Handle;
+
+  // Replaces the SystemHelper singleton with a new MockSystemHelper which is
+  // wrapped in a Handle. The MockSystemHelper can be accessed via the
+  // Handle's `mock_helper()` accessor.
+  static std::unique_ptr<Handle> replaceSystemHelper() { return std::make_unique<Handle>(); }
```

Comment:
What's the advantage to making this static and having a separate `Handle` class to contain the swapping to MockSystemHelper logic, instead of inlining that logic in SystemHelperPeer and getting rid of the `Handle` class?

Reply:
Great question! This design is a bit baroque. My thinking is that `replaceSystemHelper()` is potentially dangerous. I didn't want test A to replace the system helper and forget to put it back, thus causing problems with test B. The Handle class helps with this by putting the original helper back when it is destroyed. So as long as you have the Handle, your replacement is valid. But once you don't it goes back to normal. If you put the Handle in a local variable of your test, or member of your test, it's guaranteed not to leak.

But it's definitely baroque. I *mildly* prefer this approach, but probably only because I thought it up. I'd be happy to simplify if you think it's not worth it.

Reply:
Oh sorry, I think I wasn't clear. I definitely agree with the RAII approach with swapping back in the destructor.  I was more wondering why not just require each test to create a new SystemHelperPeer, and in the destructor of SystemHelperPeer, do the swap back to the original SystemHelper instance.  That would remove one layer of indirection w/ the `Handle` class.  But, I think your approach has the advantage that, by forcing the caller to invoke the static `replaceSystemHelper` method, it makes the semantics clear of what the SystemHelperPeer is trying to do.

I don't really have a preference either way and am happy with whichever route you prefer.  I mostly just wanted to understand the thinking behind it.  But in thinking about it a bit more, I do like the explicit nature of the replaceSystemHelper call instead of having to recreate a SystemHelperPeer each time, so I slightly prefer your current approach, even if it requires the extra `Handle` class.